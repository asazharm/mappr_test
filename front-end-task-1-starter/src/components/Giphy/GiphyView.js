import {Grid} from "@giphy/react-components";
import React, {useEffect, useState} from "react";
import {GiphyFetch} from "@giphy/js-fetch-api";
import './GiphyFetch.css'
import {BiCodeCurly} from "react-icons/bi";
import {FaRegUserCircle} from "react-icons/fa";
import {MdFavorite, MdTitle} from "react-icons/md";
import {IoMdHeartDislike} from "react-icons/io";

function GiphyView (props){
  const gf = new GiphyFetch(props.apiKey)


  const Overlay = ({ gif, isHovered }: GifOverlayProps) => {
    const [favorite, setFavorite] = useState(false);
    const [infoItemShow, setInfoItemShow] = useState('')

    useEffect(()=>{
      if(props.favorites.includes(gif.id))
        setFavorite(true)
      else
        setFavorite(false)
    })

    const handlerSetFavorite = (event) => {
      setFavorite(!favorite)
      props.setFavorite(event.currentTarget.id)
    }

    const handleInfoItemClick = (event) => {
      if (infoItemShow === event.currentTarget){
        setInfoItemShow('')
        event.currentTarget.classList.remove("show")
      }
      else if (infoItemShow !== ''){
        infoItemShow.classList.remove("show")
        event.currentTarget.className += " show"
        setInfoItemShow(event.currentTarget)
      }
      else {
        event.currentTarget.className += " show"
        setInfoItemShow(event.currentTarget)
      }
    }

    return (
      <div className="gif_overlay">
        {
          isHovered || props.isMobile ?
            <div>
              <div id={gif.id} className="gif_favorite" onClick={handlerSetFavorite}>
                {
                  props.favorites.includes(gif.id) ?
                  <IoMdHeartDislike className="gif_favorite-icon"/> :
                  <MdFavorite className="gif_favorite-icon"/>
                }

              </div>
              <div className={"gif_info"}>
                {gif.title ?
                  <div className={"gif_info-item"} onClick={handleInfoItemClick}>
                    <span>{gif.title}</span>
                    <MdTitle className={"gif_title gif_info-icon"}/>
                  </div> : ''
                }
                {gif.username ?
                  <div className={"gif_info-item"} onClick={handleInfoItemClick}>
                    <span>{gif.username}</span>
                    <FaRegUserCircle className={"gif_user gif_info-icon"}/>
                  </div> : ''
                }
                {gif.url ?
                  <div className={"gif_info-item"} onClick={handleInfoItemClick}>
                    <span><a className="link-light" href={gif.url}>{gif.url}</a></span>
                    <BiCodeCurly className={"gif_url gif_info-icon"}/>
                  </div> : ''
                }

                {/*<div className={"gif_info-item"}>*/}
                {/*  <span>{gif.create_datetime}</span>*/}
                {/*  <IoTime className={"gif_date gif_info-icon"}/>*/}
                {/*</div>*/}
              </div>
            </div>
            : ''
              }
      </div>
    )
  }

  const gifFetch = (offset: number) => {
    if (props.search.length > 0)
      return gf.search(props.search, {offset, limit: 20, sort:props.searchSortType})
    if (props.giphyViewType === 'favorites')
      return gf.gifs(props.favorites)
    if (props.giphyViewType === 'trending')
      return gf.trending({offset, limit: 20})
  }


  return (
    <Grid key={Math.floor(Math.random() * 100)} width={props.isMobile ? window.innerWidth : 1000} columns={props.isMobile ? 1 : parseInt(props.layout)} fetchGifs={gifFetch} overlay={Overlay} noLink={true}/>
  )
}

export default GiphyView