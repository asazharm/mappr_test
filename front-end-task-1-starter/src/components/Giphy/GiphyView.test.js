import React from 'react';
import renderer from 'react-test-renderer';
import GiphyView from './GiphyView';

test('GiphyView test 1', () => {
  const component = renderer.create(
    <GiphyView
      apiKey="pwZ7tgpLESvZ3AA7R8POfAs8VgTxGp7t"
      giphyViewType="trending"
      search=""
      favorites={[]}
      setFavorite={() => {}}
      isMobile={false}
      layout={3}
      searchSortType="relevant"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('GiphyView test 2', () => {
  const component = renderer.create(
    <GiphyView
      apiKey="pwZ7tgpLESvZ3AA7R8POfAs8VgTxGp7t"
      giphyViewType="favorites"
      search=""
      favorites={['KstPDy4K2OLobLQjGG']}
      setFavorite={() => {}}
      isMobile={false}
      layout={3}
      searchSortType="relevant"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('GiphyView test 3', () => {
  const component = renderer.create(
    <GiphyView
      apiKey="pwZ7tgpLESvZ3AA7R8POfAs8VgTxGp7t"
      giphyViewType="search"
      search="test"
      favorites={['KstPDy4K2OLobLQjGG']}
      setFavorite={() => {}}
      isMobile={false}
      layout={3}
      searchSortType="relevant"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('GiphyView test mobile 1', () => {
  const component = renderer.create(
    <GiphyView
      apiKey="pwZ7tgpLESvZ3AA7R8POfAs8VgTxGp7t"
      giphyViewType="trending"
      search=""
      favorites={['KstPDy4K2OLobLQjGG']}
      setFavorite={() => {}}
      isMobile
      layout={3}
      searchSortType="relevant"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('GiphyView test mobile 2', () => {
  const component = renderer.create(
    <GiphyView
      apiKey="pwZ7tgpLESvZ3AA7R8POfAs8VgTxGp7t"
      giphyViewType="search"
      search="test"
      favorites={['KstPDy4K2OLobLQjGG']}
      setFavorite={() => {}}
      isMobile
      layout={3}
      searchSortType="relevant"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
