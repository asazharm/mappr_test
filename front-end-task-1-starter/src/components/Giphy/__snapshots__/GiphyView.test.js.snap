// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GiphyView test 1 1`] = `
<div
  className="giphy-grid"
  style={
    Object {
      "width": 1000,
    }
  }
>
  <div
    style={
      Object {
        "height": "-6px",
        "position": "relative",
        "width": "999px",
      }
    }
  />
  <div>
    <div
      className="css-7fqhxk"
    >
      <div
        className="loader css-1qt545t"
      >
        <div
          className="css-1qyiwug"
          color="#00ff99"
        />
        <div
          className="css-17jbzkf"
          color="#00ccff"
        />
        <div
          className="css-1mksb1v"
          color="#9933ff"
        />
        <div
          className="css-b6jo2y"
          color="#ff6666"
        />
        <div
          className="css-ea8625"
          color="#fff35c"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`GiphyView test 2 1`] = `
<div
  className="giphy-grid"
  style={
    Object {
      "width": 1000,
    }
  }
>
  <div
    style={
      Object {
        "height": "-6px",
        "position": "relative",
        "width": "999px",
      }
    }
  />
  <div>
    <div
      className="css-7fqhxk"
    >
      <div
        className="loader css-1qt545t"
      >
        <div
          className="css-1qyiwug"
          color="#00ff99"
        />
        <div
          className="css-17jbzkf"
          color="#00ccff"
        />
        <div
          className="css-1mksb1v"
          color="#9933ff"
        />
        <div
          className="css-b6jo2y"
          color="#ff6666"
        />
        <div
          className="css-ea8625"
          color="#fff35c"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`GiphyView test 3 1`] = `
<div
  className="giphy-grid"
  style={
    Object {
      "width": 1000,
    }
  }
>
  <div
    style={
      Object {
        "height": "-6px",
        "position": "relative",
        "width": "999px",
      }
    }
  />
  <div>
    <div
      className="css-7fqhxk"
    >
      <div
        className="loader css-1qt545t"
      >
        <div
          className="css-1qyiwug"
          color="#00ff99"
        />
        <div
          className="css-17jbzkf"
          color="#00ccff"
        />
        <div
          className="css-1mksb1v"
          color="#9933ff"
        />
        <div
          className="css-b6jo2y"
          color="#ff6666"
        />
        <div
          className="css-ea8625"
          color="#fff35c"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`GiphyView test mobile 1 1`] = `
<div
  className="giphy-grid"
  style={
    Object {
      "width": 1024,
    }
  }
>
  <div
    style={
      Object {
        "height": "-6px",
        "position": "relative",
        "width": "1024px",
      }
    }
  />
  <div>
    <div
      className="css-7fqhxk"
    >
      <div
        className="loader css-1qt545t"
      >
        <div
          className="css-1qyiwug"
          color="#00ff99"
        />
        <div
          className="css-17jbzkf"
          color="#00ccff"
        />
        <div
          className="css-1mksb1v"
          color="#9933ff"
        />
        <div
          className="css-b6jo2y"
          color="#ff6666"
        />
        <div
          className="css-ea8625"
          color="#fff35c"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`GiphyView test mobile 2 1`] = `
<div
  className="giphy-grid"
  style={
    Object {
      "width": 1024,
    }
  }
>
  <div
    style={
      Object {
        "height": "-6px",
        "position": "relative",
        "width": "1024px",
      }
    }
  />
  <div>
    <div
      className="css-7fqhxk"
    >
      <div
        className="loader css-1qt545t"
      >
        <div
          className="css-1qyiwug"
          color="#00ff99"
        />
        <div
          className="css-17jbzkf"
          color="#00ccff"
        />
        <div
          className="css-1mksb1v"
          color="#9933ff"
        />
        <div
          className="css-b6jo2y"
          color="#ff6666"
        />
        <div
          className="css-ea8625"
          color="#fff35c"
        />
      </div>
    </div>
  </div>
</div>
`;
